src/helics/core/mpi/MpiService.cpp

```diff
@@ -0,0 +1,314 @@
+
+*/
+
+#include "MpiService.h"
+
+namespace helics {
+
+MPI_Comm MpiService::mpiCommunicator = MPI_COMM_NULL;
+
+MpiService& MpiService::getInstance ()
+{
+    static MpiService instance;
+    instance.startService ();
+    return instance;
+}
+
+void MpiService::setMpiCommunicator (MPI_Comm communicator)
+{
+    mpiCommunicator = communicator;
+}
+
+MpiService::MpiService ()
+    : commRank (-1)
+{
+    comms_connected = 0;
+    stop_service = false;
+    startup_flag = false;
+}
+
+MpiService::~MpiService ()
+{
+    // Stop the service thread
+    stop_service = true;
+    if (service_thread->joinable ())
+    {
+        service_thread->join ();
+    }
+}
+
+void MpiService::startService ()
+{
+    if (service_thread == nullptr)
+    {
+        startup_flag = true;
+        service_thread = std::make_unique<std::thread> (&MpiService::serviceLoop, this);
+        while (startup_flag && !stop_service);
+    }
+}
+
+void MpiService::serviceLoop ()
+{
+    // Startup/teardown for MPI can all go in separate functions
+    // For integrating helics MPI calls with existing MPI applications, user may want to call from their own thread
+
+    // Initialize MPI
+    if (initMPI ())
+    {
+        // if mpiCommunicator isn't set by user, make it a duplicate of MPI_COMM_WORLD
+        if (mpiCommunicator == MPI_COMM_NULL)
+        {
+            MPI_Comm_dup (MPI_COMM_WORLD, &mpiCommunicator);
+        }
+
+        // set commRank to our process rank
+        MPI_Comm_rank (mpiCommunicator, &commRank);
+    }
+
+    // signal that we have finished starting
+    startup_flag = false;
+
+    std::cout << "Started MPI service loop for rank " << commRank << std::endl;
+
+
+    // Run as long as we have something in the send queue or the chance of getting something
+    while (!stop_service || comms_connected > 0 || !(txMessageQueue.empty ()))
+    {
+        // send/receive MPI messages
+        sendAndReceiveMessages ();
+    }
+
+    // Post receives for any waiting sends
+    int message_waiting = 1;
+    MPI_Status status;
+    while (message_waiting)
+    {
+        MPI_Iprobe (MPI_ANY_SOURCE, MPI_ANY_TAG, mpiCommunicator, &message_waiting, &status);
+        if (message_waiting)
+        {
+            // Get the size of the message waiting to be received
+            int recv_size;
+            std::vector<char> buffer;
+            MPI_Get_count (&status, MPI_CHAR, &recv_size);
+            buffer.resize (recv_size);
+            
+            // Receive the message
+            MPI_Recv (buffer.data (), buffer.capacity (), MPI_CHAR, MPI_ANY_SOURCE, MPI_ANY_TAG, mpiCommunicator, &status);
+        }
+    }
+
+    // Finalize MPI
+    int mpi_initialized;
+    MPI_Initialized(&mpi_initialized);
+    
```

Comment:
This check is okay, but if MPI_Finalize() was called outside of HELICS, then the above MPI_IProbe and MPI_Recv calls would likely error.

Most programs I am aware of will call MPI_Init() right inside their main(), and then MPI_Finalize() right before returning from main().  Is there any danger in HELICS calling MPI_Finalize() here?  Would it mess up an app that is also using MPI?

Reply:
It is possible, but if they're using MPI in their own program, they will want to call the sendAndReceiveMessages function as part of their own loop instead of letting HELICS create a thread (everything below MPI_THREAD_SERIALIZED wants the MPI calls to happen on the same thread as MPI_Init and MPI_Finalized get called by), or a mutex would need to get added to serialize MPI calls unless their MPI implementation has full threading support.

Reply:
I split the part that is posting receives during shutdown into its own public method, and added a check when initializing MPI to see if the application had already initialized MPI -- now when the thread is closing it will only finalize MPI if it also initialized MPI. If they're trying to use the MPI core with anything less than MPI_THREAD_MULTIPLE, they can call setStartServiceThread(false) to make so that helics doesn't spawn the thread to make MPI calls.